File: src/lib/vault-core/contracts/results.ts
/**
 * FEAT-003 vault-core contracts — closed typed result union.
 *
 * Expected adapter failures return a closed discriminated result union with safe
 * structured fields: stable typed code, retryable flag, allowed recovery actions,
 * optional bounded retry deadline, optional random per-occurrence support code.
 * No raw exception, path, key alias, full identity, ciphertext excerpt, or free-form
 * platform message crosses the boundary.
 *
 * Deep-Dive override: `NetworkMismatch` is NOT an emitted v1 behavior; future network
 * binding requires a new version. v1 registry: 18 codes.
 *
 * Normative source: FEAT-003 FeatureDescription "Typed Result Contract".
 */

/** Allowed recovery actions surfaced to FEAT-002 for safe UX decisions. */
export type AllowedRecoveryAction =
  | 'retry'
  | 'reprovision'
  | 'verifyOnline'
  | 'requestPersistence'
  | 'unlockPlatformProtection'
  | 'changePassword'
  | 'clearRemovalTombstone'
  | 'resumeRemoval';

/** Closed v1 result codes (18). `NetworkMismatch` reserved for a future version. */
export type VaultResultCode =
  | 'NoVault'
  | 'UnsupportedVaultVersion'
  | 'MalformedEnvelope'
  | 'WrongPasswordOrDamagedData'
  | 'Throttled'
  | 'KdfResourceLimit'
  | 'PlatformProtectionUnavailable'
  | 'PlatformProtectionInvalidated'
  | 'IdentityBindingMismatch'
  | 'MigrationFailedRollbackAvailable'
  | 'GenerationConflict'
  | 'StorageUnavailable'
  | 'StorageQuotaExceeded'
  | 'PersistenceDenied'
  | 'StaleSession'
  | 'OperationForbidden'
  | 'CleanupFailed'
  | 'ExtensionUnsupported';

/** Closed registry: every v1 code with its safe metadata (declarative, exhaustive). */
export interface ResultCodeMeta {
  readonly code: VaultResultCode;
  readonly retryable: boolean;
  readonly allowedActions: readonly AllowedRecoveryAction[];
}

export const VAULT_RESULT_CODES: readonly VaultResultCode[] = [
  'NoVault',
  'UnsupportedVaultVersion',
  'MalformedEnvelope',
  'WrongPasswordOrDamagedData',
  'Throttled',
  'KdfResourceLimit',
  'PlatformProtectionUnavailable',
  'PlatformProtectionInvalidated',
  'IdentityBindingMismatch',
  'MigrationFailedRollbackAvailable',
  'GenerationConflict',
  'StorageUnavailable',
  'StorageQuotaExceeded',
  'PersistenceDenied',
  'StaleSession',
  'OperationForbidden',
  'CleanupFailed',
  'ExtensionUnsupported',
] as const;

export const VAULT_RESULT_REGISTRY: Readonly<Record<VaultResultCode, ResultCodeMeta>> = {
  NoVault: { code: 'NoVault', retryable: false, allowedActions: ['reprovision'] },
  UnsupportedVaultVersion: { code: 'UnsupportedVaultVersion', retryable: false, allowedActions: [] },
  MalformedEnvelope: { code: 'MalformedEnvelope', retryable: true, allowedActions: ['reprovision'] },
  WrongPasswordOrDamagedData: { code: 'WrongPasswordOrDamagedData', retryable: true, allowedActions: ['retry', 'reprovision'] },
  Throttled: { code: 'Throttled', retryable: true, allowedActions: ['retry'] },
  KdfResourceLimit: { code: 'KdfResourceLimit', retryable: true, allowedActions: ['verifyOnline'] },
  PlatformProtectionUnavailable: { code: 'PlatformProtectionUnavailable', retryable: true, allowedActions: ['unlockPlatformProtection', 'retry'] },
  PlatformProtectionInvalidated: { code: 'PlatformProtectionInvalidated', retryable: false, allowedActions: ['reprovision'] },
  IdentityBindingMismatch: { code: 'IdentityBindingMismatch', retryable: false, allowedActions: ['reprovision'] },
  MigrationFailedRollbackAvailable: { code: 'MigrationFailedRollbackAvailable', retryable: true, allowedActions: ['retry'] },
  GenerationConflict: { code: 'GenerationConflict', retryable: true, allowedActions: ['retry'] },
  StorageUnavailable: { code: 'StorageUnavailable', retryable: true, allowedActions: ['retry'] },
  StorageQuotaExceeded: { code: 'StorageQuotaExceeded', retryable: true, allowedActions: ['requestPersistence'] },
  PersistenceDenied: { code: 'PersistenceDenied', retryable: true, allowedActions: ['requestPersistence'] },
  StaleSession: { code: 'StaleSession', retryable: true, allowedActions: ['retry'] },
  OperationForbidden: { code: 'OperationForbidden', retryable: false, allowedActions: [] },
  CleanupFailed: { code: 'CleanupFailed', retryable: true, allowedActions: ['retry', 'resumeRemoval'] },
  ExtensionUnsupported: { code: 'ExtensionUnsupported', retryable: false, allowedActions: [] },
} as const;

/** Random per-occurrence support code (bounded, safe for local diagnostics). */
export type SupportCode = string;

/** One typed vault failure — the only shape crossing the core boundary on expected failure. */
export interface VaultFailure {
  readonly ok: false;
  readonly code: VaultResultCode;
  readonly retryable: boolean;
  readonly allowedActions: readonly AllowedRecoveryAction[];
  /** Optional bounded retry deadline (epoch ms); present when retryable with a cooldown. */
  readonly retryDeadlineMs?: number;
  /** Optional random per-occurrence support code for sanitized local diagnostics. */
  readonly supportCode?: SupportCode;
}

/** One typed vault success — closed shape with only safe public fields. */
export interface VaultSuccess<T> {
  readonly ok: true;
  readonly value: T;
}

/** Closed discriminated result union. */
export type VaultResult<T> = VaultSuccess<T> | VaultFailure;

/** Construct a safe typed failure from the closed registry. */
export function failure(
  code: VaultResultCode,
  extra: { readonly retryDeadlineMs?: number; readonly supportCode?: SupportCode } = {},
): VaultFailure {
  const meta = VAULT_RESULT_REGISTRY[code];
  return {
    ok: false,
    code,
    retryable: meta.retryable,
    allowedActions: meta.allowedActions,
    ...(extra.retryDeadlineMs !== undefined ? { retryDeadlineMs: extra.retryDeadlineMs } : {}),
    ...(extra.supportCode !== undefined ? { supportCode: extra.supportCode } : {}),
  };
}

/** Construct a safe typed success. */
export function success<T>(value: T): VaultSuccess<T> {
  return { ok: true, value };
}

/** Exhaustive registry check used by tests: every code has exactly one meta entry. */
export function assertRegistryExhaustive(): void {
  const registered = Object.keys(VAULT_RESULT_REGISTRY).sort();
  const declared = [...VAULT_RESULT_CODES].sort();
  if (JSON.stringify(registered) !== JSON.stringify(declared)) {
    throw new Error('vault result registry is not exhaustive: codes must match VAULT_RESULT_CODES');
  }
}

File: src-tauri/src/ubuntu_vault/contracts/results.rs
//! Closed native result union (FEAT-005 "Error Handling").
//!
//! Every expected Rust, D-Bus, filesystem, TLS/gRPC, crypto, and IPC failure
//! maps to this closed union. Safe platform subreason may choose UI recovery
//! but never exposes raw provider/path details. This mirrors FEAT-003's
//! typed-result philosophy in native code.

use serde::{Deserialize, Serialize};

/// Recovery actions the UI may offer for a typed failure.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum RecoveryAction {
    Retry,
    Reprovision,
    VerifyOnline,
    UnlockPlatformProtection,
    EnableOsProtection,
    PortableRecovery,
    ClearRemovalTombstone,
    ResumeRemoval,
    Cancel,
}

/// Closed native error codes. Mapping to FEAT-003's vault result codes happens
/// in the bridge layer; native codes here are the adapter's own vocabulary.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum NativeErrorCode {
    NoVault,
    UnsupportedVaultVersion,
    MalformedEnvelope,
    WrongPasswordOrDamagedData,
    Throttled,
    KdfResourceLimit,
    PlatformProtectionUnavailable,
    PlatformProtectionInvalidated,
    IdentityBindingMismatch,
    MigrationFailedRollbackAvailable,
    GenerationConflict,
    StorageUnavailable,
    StorageQuotaExceeded,
    PersistenceDenied,
    StaleSession,
    OperationForbidden,
    CleanupFailed,
    ExtensionUnsupported,
    ProviderAbsent,
    ProviderLocked,
    PromptCancelled,
    PromptTimedOut,
    ProviderTemporarilyUnavailable,
    UnqualifiedProvider,
    WrapperAmbiguous,
    WrapperVersionUnsupported,
    BuildVersionMismatch,
    NetworkTimeout,
    ProfileNotFound,
    KeyMismatch,
    RemovalIncomplete,
}

impl NativeErrorCode {
    /// Safe recovery action set. No raw detail is ever included.
    pub fn recovery_actions(self) -> &'static [RecoveryAction] {
        match self {
            Self::NoVault => &[RecoveryAction::Reprovision],
            Self::UnsupportedVaultVersion | Self::MalformedEnvelope => &[RecoveryAction::Cancel],
            Self::WrongPasswordOrDamagedData => {
                &[RecoveryAction::Retry, RecoveryAction::Reprovision]
            }
            Self::Throttled => &[RecoveryAction::Retry],
            Self::KdfResourceLimit => &[RecoveryAction::Retry],
            Self::PlatformProtectionUnavailable => &[
                RecoveryAction::Retry,
                RecoveryAction::UnlockPlatformProtection,
            ],
            Self::PlatformProtectionInvalidated => &[RecoveryAction::PortableRecovery],
            Self::IdentityBindingMismatch | Self::KeyMismatch => &[
                RecoveryAction::Retry,
                RecoveryAction::VerifyOnline,
                RecoveryAction::Reprovision,
            ],
            Self::MigrationFailedRollbackAvailable => &[RecoveryAction::Retry],
            Self::GenerationConflict => &[RecoveryAction::Retry, RecoveryAction::Cancel],
            Self::StorageUnavailable | Self::StorageQuotaExceeded => &[RecoveryAction::Retry],
            Self::PersistenceDenied => &[RecoveryAction::Cancel],
            Self::StaleSession => &[RecoveryAction::Retry],
            Self::OperationForbidden => &[RecoveryAction::Cancel],
            Self::CleanupFailed => &[RecoveryAction::Retry],
            Self::ExtensionUnsupported => &[RecoveryAction::Cancel],
            Self::ProviderAbsent => &[
                RecoveryAction::Retry,
                RecoveryAction::EnableOsProtection,
                RecoveryAction::PortableRecovery,
            ],
            Self::ProviderLocked | Self::PromptCancelled | Self::PromptTimedOut => &[
                RecoveryAction::Retry,
                RecoveryAction::UnlockPlatformProtection,
            ],
            Self::ProviderTemporarilyUnavailable => &[RecoveryAction::Retry],
            Self::UnqualifiedProvider => &[RecoveryAction::EnableOsProtection],
            Self::WrapperAmbiguous => &[RecoveryAction::PortableRecovery],
            Self::WrapperVersionUnsupported | Self::BuildVersionMismatch => {
                &[RecoveryAction::Cancel]
            }
            Self::NetworkTimeout => &[RecoveryAction::Retry],
            Self::ProfileNotFound => &[RecoveryAction::VerifyOnline, RecoveryAction::Cancel],
            Self::RemovalIncomplete => &[RecoveryAction::Retry, RecoveryAction::ResumeRemoval],
        }
    }
}

/// A safe native outcome: success carries a typed payload kind; failure
/// carries a closed code plus optional support code. No raw detail.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", tag = "outcome", deny_unknown_fields)]
pub enum NativeOutcome {
    Ok {
        kind: OutcomeKind,
    },
    Err {
        code: NativeErrorCode,
        support_code: Option<u64>,
    },
}

/// Safe success payload kinds (opaque handles or coarse state only).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum OutcomeKind {
    Locked,
    Unlocked,
    Provisioned,
    Verified,
    Removed,
    Preview,
    RevealPrepared,
    Signed,
    DatImported,
    DatExported,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn error_codes_have_closed_recovery_actions() {
        // Every code maps to a non-empty closed action set.
        let codes = [
            NativeErrorCode::NoVault,
            NativeErrorCode::UnsupportedVaultVersion,
            NativeErrorCode::MalformedEnvelope,
            NativeErrorCode::WrongPasswordOrDamagedData,
            NativeErrorCode::Throttled,
            NativeErrorCode::KdfResourceLimit,
            NativeErrorCode::PlatformProtectionUnavailable,
            NativeErrorCode::PlatformProtectionInvalidated,
            NativeErrorCode::IdentityBindingMismatch,
            NativeErrorCode::MigrationFailedRollbackAvailable,
            NativeErrorCode::GenerationConflict,
            NativeErrorCode::StorageUnavailable,
            NativeErrorCode::StorageQuotaExceeded,
            NativeErrorCode::PersistenceDenied,
            NativeErrorCode::StaleSession,
            NativeErrorCode::OperationForbidden,
            NativeErrorCode::CleanupFailed,
            NativeErrorCode::ExtensionUnsupported,
            NativeErrorCode::ProviderAbsent,
            NativeErrorCode::ProviderLocked,
            NativeErrorCode::PromptCancelled,
            NativeErrorCode::PromptTimedOut,
            NativeErrorCode::ProviderTemporarilyUnavailable,
            NativeErrorCode::UnqualifiedProvider,
            NativeErrorCode::WrapperAmbiguous,
            NativeErrorCode::WrapperVersionUnsupported,
            NativeErrorCode::BuildVersionMismatch,
            NativeErrorCode::NetworkTimeout,
            NativeErrorCode::ProfileNotFound,
            NativeErrorCode::KeyMismatch,
            NativeErrorCode::RemovalIncomplete,
        ];
        for code in codes {
            assert!(!code.recovery_actions().is_empty(), "{code:?}");
        }
    }

    #[test]
    fn invalidated_never_offers_retry_replacement() {
        let actions = NativeErrorCode::PlatformProtectionInvalidated.recovery_actions();
        assert!(actions.contains(&RecoveryAction::PortableRecovery));
        assert!(!actions.contains(&RecoveryAction::Retry));
    }

    #[test]
    fn outcome_round_trips_closed() {
        let err = NativeOutcome::Err {
            code: NativeErrorCode::WrongPasswordOrDamagedData,
            support_code: Some(1234),
        };
        let json = serde_json::to_string(&err).unwrap();
        let back: NativeOutcome = serde_json::from_str(&json).unwrap();
        assert_eq!(err, back);
        assert!(!json.contains("password"));
    }

    #[test]
    fn unknown_fields_are_rejected_fail_closed() {
        // A malformed, unsupported, or stale request must fail closed without
        // changing keyring or vault state — unknown fields never deserialize.
        let json = r#"{"outcome":"err","code":"wrongPasswordOrDamagedData","supportCode":1234,"rawDetail":"secret"}"#;
        assert!(serde_json::from_str::<NativeOutcome>(json).is_err());
    }

    #[test]
    fn support_code_is_bounded() {
        // Support codes are random per-occurrence integers only; no raw detail.
        let err = NativeOutcome::Err {
            code: NativeErrorCode::CleanupFailed,
            support_code: Some(u64::MAX),
        };
        let json = serde_json::to_string(&err).unwrap();
        assert!(json.contains("18446744073709551615"));
        assert!(!json.contains("path") && !json.contains("dbus"));
    }
}
